lib/service/service.go

```diff
@@ -4813,3 +4822,73 @@ func isDebugMode() bool {
 	v, _ := strconv.ParseBool(os.Getenv(teleport.DebugEnvVar))
 	return v
 }
+
+// readHostUUUIDFromStorage tries to read the `host_uuid` value from different storages,
+// depending on where the process is running.
+// If the process is running in a Kubernetes Cluster, this function will attempt
+// to read the `host_uuid` from the Kubernetes Secret. If it does not exist or
+// if it is not running on a Kubernetes cluster the read is done from the local
+// storage: `dataDir/host_uuid`.
+func readHostUUIDFromStorages(ctx context.Context, dataDir string) (string, error) {
+	if kubernetes.InKubeCluster() {
+		if hostUUID, err := loadHostUUIDFromKubeSecret(ctx); err == nil && len(string(hostUUID)) > 0 {
+			return string(hostUUID), nil
+		}
+	}
+	// Even if running in Kubernetes fallback to local storage if `host_uuid` was
+	// not found in secret.
+	hostUUID, err := utils.ReadHostUUID(dataDir)
+	return hostUUID, trace.Wrap(err)
+}
+
+// persistHostUUIDToStorages writes the HostUUID to local data and  and to
+// Kubernetes Secret if this process is running on a Kubernetes Cluster.
+func persistHostUUIDToStorages(ctx context.Context, cfg *Config) error {
+	if err := utils.WriteHostUUID(cfg.DataDir, cfg.HostUUID); err != nil {
```

Comment:
When running inside Kubernetes, why does it need to be saved to filesystem? `readHostUUIDFromStorages` will still read it from a secret, no?

Reply:
I reckon that `tctl` might need the file on disk to work properly - although that would only apply to the auth server storing its own credentials in kubernetes, which isn't really needed.

Reply:
Kube secret identity storage only applies to the Kube Agent chart which does not have the auth server.
I wrote the `host_uuid` to disk for the cases where storage is enabled and we need to downgrade to a pre-11 version